---
title:      JSON.stringify()报错处理
date:       2020-07-02
categories:
    - 知识整理
tags:
    - 前端
    - JSON
    - 报错处理
excerpt: 有待解决的问题
---
# JSON.stringify()报错处理

## 报错信息

```

Converting circular structure to JSON；

```

## 解析

我们使用JSON.parse想要解析的对象中存在循环引用如下：

```javascript

var a = {};

a.b = a;

```

明显我们想为b属性赋值的时候直接指定了a本身，也就是指向了a的内存地址，而并非只是a的值所以形成了一个循环，既然如此解决方法也就呼之欲出，我们使用深复制造多一个a就好了，既然我们这里也使用了JSON，那直接使用JSON来深拷贝就好了，我们的数据里也不会有方法需要拷贝：

```javascript

var a = {};

a.b = JSON.parse(JSON.stringify(a));

```

## 实际情况

但是实际上笔者想要实现的是在全局路由守卫中把from的信息存到sessionStorage中，所以需要JSON的序列化和反序列化。报出了上面这个错误，明显并不是自己写的对象循环引用，估计是Vue router的问题，报错如下：

```

starting at object with constructor 'Vue'

 property '$options' -> object with constructor 'Object'

 property 'router' -> object with constructor 'VueRouter'

 --- property 'app' closes the circle

```

既然不是自己写的那上面的JSON深复制方法自然也无法解决我们的问题，虽然笔者直接取到from里的path属性解决了这个问题，但是难免以后会需要真的存入一个对象，还是得弄清楚怎么回事。虽然报错是Vue router，但如果单纯使用vue Router应该并不会引起循环调用，应该是在别处还对this.$router有调用，仔细看看项目，我们的路由信息是由后台提供的，获取的路由信息会被存到vuex里面作状态管理，那很可能路由还被vuex调用了，我们解决思路仍然是进行深复制，不过这次要进行彻底的递归深复制，准备一个深复制方法：

```javascript

export const deepClone = (obj) => {

  if (typeof obj !== 'object') {

    return null

  }



  let copy = isArray(obj) ? [] : {}



  for (let key in obj) {

    let item

    if (isArray(copy)) {

      item = obj[key] || ''

    } else if (isObject(copy)) {

      if (obj.hasOwnProperty(key)) {

        item = obj[key] || ''

      }

    }

    copy[key] = typeof item === 'object' ? deepClone(item) : item

  }



  return copy

}

```

我们用这个方法去给from对象深复制一遍，发现还是报错...

```

Uncaught TypeError: obj.hasOwnProperty is not a function

```

很奇怪，明明传进去的from是对象类型为什么会没有这个原生方法呢？查阅一些资料发现from有可能是通过Object.create(null)创建的，这样跟Object的原型链就会脱节，所以报错会不存在这个方法。解决方法就是用call调用原型链顶部的Object方法：

```javascript

// 把obj.hasOwnProperty写成下面这个

Object.prototype.hasOwnProperty.call(obj, key)

```

本来以为问题能顺利解决了，然而还是报错...

```

RangeError: Maximum call stack size exceeded     at Vue.hasOwnProperty (<anonymous>)

```

报错信息明显是深复制里的递归死循环了，思考了一下为什么会死循环，发现有可能是因为这个对象本来就是个重复引用的对象所以我们在检查hasOwnProperty的时候自然也是死循环。。。那这个问题本身也形成了个死循环。。。只能说下来再有待解决了

## 参考链接

[深拷贝1](https://www.cnblogs.com/duanlibo/p/11609716.html)

[深拷贝2](https://www.cnblogs.com/hyns/p/12405328.html)

[hasOwnProperty失效](https://www.jianshu.com/p/e7b21271877d)

[JSON报错分析1](https://blog.csdn.net/lydxwj/article/details/103489794)

[JSON报错分析2](https://stackoverflow.com/questions/4816099/chrome-sendrequest-error-typeerror-converting-circular-structure-to-json)
